lambdas/cdk/init-stack.ts

```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env node
+/* eslint-disable no-new */
+import * as cdk from '@aws-cdk/core';
+import ContactCompleteStack from './contact-complete-stack';
+import ContactCoreStack from './contact-core-stack';
+import ContactRetrieveStack from './contact-retrieve-stack';
+
+/**
```

Comment:
I tend to prefer to include "discussion" tagged comments that I don't actually expect to be acted on in the near term in the code. If they are PR comments they tend to get lost in the shuffle. If they live in the code, the planning and reasoning follows the code a little more closely. 

When someone comes back in a month or two and thinks "why is it like this" they can see immediately that there are ongoing plans and discussions. When the comments are signed they even know who to discuss the topic with if they have ideas. It is basically a place for ideas/discussion between "PR Comment" and "Create a Ticket".

It boils down to this: if it is a discussion that I expect to be completely resolved in the span of the PR, I try to make it a PR comment. If I expect it to last longer, I tend to add a comment in the code.

I'm, obviously, happy to defer to the team's preference.

Reply:
My preference would be to raise the discussion topics as PR comments, have the discussion, and have any future work arising from the discussion added as TODO items in the code and / or backlog tickets (even if that ticket is just 'investigate the best option').

Or if it's a very far future item maybe just a brief 1 liner identifying the point that needs further thinking. I just find that large blocks of prose in comments (other than specific 'How to' information) tends to lose relevance and / or be ignored not long after it's initially written anyway, and after that it's just clutter.

We can see what the team thinks though

Reply:
As I'm about to create another discussion comment for the next PR (sorry. habits die hard :)), I realize there is another reason I like them, especially in code like this that includes a lot of new architectural decisions: They allow me to ask questions and outline my thought process about future plans and implementation details while I'm working on a problem and actively thinking about it without blocking myself to write a slack question that may take a while to discuss asynchronously and without taking notes somewhere else for later use in a PR comment. 

Going back after the fact to add a PR comment requires keeping notes somewhere outside of code while the problem is fresh until I'm ready to open a PR, remembering to look at those notes, working out what code the note entries are referring to, and then rewording them to make sense in the context of a PR comment. For me, inline discussion/decision comments significantly reduce the brain cycles required to document and discuss architectural decisions. Throw in the fact that most IDEs include the ability (or an extension) to collapse long comments by default if an individual dev finds them distracting, and I have trouble seeing a real downside so long as they are relevant and potentially useful in the future to understanding why something is the way it is.

Reply:
It sounds like you could comment inline and extract them out when it comes time to write up a PR if that helps you? To me the inline comments are still clutter the majority of the time anybody working with the code is looking at it. 

I'm trying to save brain cycles of people working on the code (as opposed to discussing architecture). Sure, you can collapse them in an IDE, but they are still there as a distraction, and we aren't always looking at code in an IDE. Small files are much easier to work with in PRs.

To me, inline comments in code should be brief and on point, directly relevant of somebody trying to understand the code they are reading. Grokking code can be tough enough, my thinking is that we should make every effort to reduce that cognitive load where we can.

I suppose it's not _as_ disruptive if the discussion comments are always at the top, but I would still find them a distraction. The last thing I want to do is inadvertently train my brain to start filtering out comment blocks in the code I'm looking at.

Also, for architectural discussions, having them tightly coupled to a specific commit isn't of much value, since they tend to range broader than a single branch / source file. Just having the date of updates is context enough, and everything has that in some form (Jira and Box have version histories, PRs date every contribution etc.)

Links to .md files elsewhere in the source seem like a fair compromise to me. As a dev I like stuff to be in code repos because that's where I spend my time - I just prefer it not to be in the same files as the source I'm working on.

Reply:
Pulling my more "discussion" type comments out at PR creation time and moving longer decision/planning overview into a docs folder with small inline links seems like a great compromise to me.

Reply:
After a little bit more thought, I will note that I think the comment standard is likely different for "code" and "IaC". This specific example that we are talking about is a little weird because it is javascript, but it is really IaC. I think that in IaC it is more important to have some of the reasoning behind the decision that was made and possible options moving forward inside of the code since, generally, the "code" isn't really doing anything all that complex and the promise that IaC is "self documenting" isn't always true.

I also think this example is a little more "prose-like" than I would normally put in these types of comments because of the passive tone I'm using.